src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -794,9 +798,7 @@ public ClusterState execute(ClusterState currentState) {
                         }
 
                         ClusterState updatedState = stateToProcess.clusterState;
-
-                        // if the new state has a smaller version, and it has the same master node, then no need to process it
-                        if (updatedState.version() < currentState.version() && Objects.equal(updatedState.nodes().masterNodeId(), currentState.nodes().masterNodeId())) {
+                        if (newClusterStateDated(currentState, updatedState)) {
```

Comment:
can we extract this logic and the logic above (where we prune/ dedup incoming cluster states), into a static method and unit test it? would make me sleep better :)